### PR TITLE
Vpc cluster worker volume

### PR DESCRIPTION
**Pull request by Anil-CM**
Screenshot 2021-01-28 at 6 03 09 PM
Support for VPC Volume attachment to VPC Cluster worker node.

### CODE REVIEW

File: ibm/data_source_ibm_container_vpc_cluster_worker_volume_attachment.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func dataSourceIBMContainerVpcWorkerVolumeAttachment() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceIBMContainerVpcWorkerVolumeAttachmentRead,
+
+		Schema: map[string]*schema.Schema{
+			"volume_attachament": {
+				Type:        schema.TypeString,
```
Comment (hkantare) [inline ibm/data_source_ibm_container_vpc_cluster_worker_volume_attachment.go:13]: change description to The volume attachemnt ID

File: ibm/provider.go [hunk 2]
```diff
@@ -154,117 +154,118 @@ func Provider() terraform.ResourceProvider {
...
+			"ibm_container_bind_service":                 dataSourceIBMContainerBindService(),
+			"ibm_container_cluster":                      dataSourceIBMContainerCluster(),
+			"ibm_container_cluster_config":               dataSourceIBMContainerClusterConfig(),
+			"ibm_container_cluster_versions":             dataSourceIBMContainerClusterVersions(),
+			"ibm_container_cluster_worker":               dataSourceIBMContainerClusterWorker(),
+			"ibm_container_vpc_cluster_alb":              dataSourceIBMContainerVPCClusterALB(),
+			"ibm_container_vpc_alb":                      dataSourceIBMContainerVPCClusterALB(),
+			"ibm_container_vpc_cluster":                  dataSourceIBMContainerVPCCluster(),
+			"ibm_container_vpc_cluster_worker":           dataSourceIBMContainerVPCClusterWorker(),
+			"ibm_container_vpc_worker_volume_attachment": dataSourceIBMContainerVpcWorkerVolumeAttachment(),
```
Comment (hkantare) [inline ibm/provider.go:202]: can we rename ibm_container_vpc_worker_volume_attachment to ibm_container_vpc_worker_storage

File: ibm/provider.go [hunk 3]
```diff
@@ -356,6 +357,7 @@ func Provider() terraform.ResourceProvider {
 			"ibm_container_vpc_alb":                              resourceIBMContainerVpcALB(),
 			"ibm_container_vpc_worker_pool":                      resourceIBMContainerVpcWorkerPool(),
 			"ibm_container_vpc_cluster":                          resourceIBMContainerVpcCluster(),
+			"ibm_container_vpc_worker_volume_attachment":         resourceIBMContainerVpcWorkerVolumeAttachment(),
```
Comment (hkantare) [inline ibm/provider.go:360]: can we rename ibm_container_vpc_worker_volume_attachment to ibm_container_vpc_worker_storage

File: ibm/resource_ibm_container_vpc_cluster_worker_volume_attachment.go [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+
+			"cluster": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Cluster name or ID",
+			},
+
+			"worker": {
+				Type:        schema.TypeString,
```
Comment (hkantare) [inline ibm/resource_ibm_container_vpc_cluster_worker_volume_attachment.go:48]: Add ForceNew here also

File: ibm/resource_ibm_container_vpc_cluster_worker_volume_attachment.go [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+	if err != nil {
+		fmt.Println(err)
+		return err
+	}
+	d.SetId(volumeattached.Id)
+	_, attachErr := waitforVolumetoAttach(d, meta)
+	if attachErr != nil {
+		return attachErr
+	}
+	return resourceIBMContainerVpcWorkerVolumeAttachmentUpdate(d, meta)
```
Comment (hkantare) [inline ibm/resource_ibm_container_vpc_cluster_worker_volume_attachment.go:106]: return read

File: ibm/resource_ibm_container_vpc_cluster_worker_volume_attachment.go [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+		VolumeID: volumeID,
+		Worker:   workerID,
+	}
+
+	volumeattached, err := workersAPI.CreateStorageAttachment(attachVolumeRequest, target)
+	if err != nil {
+		fmt.Println(err)
+		return err
+	}
+	d.SetId(volumeattached.Id)
```
Comment (hkantare) [inline ibm/resource_ibm_container_vpc_cluster_worker_volume_attachment.go:101]: Need to set ID combination of cluster/workerid/volumeattacheemntID

Comment (hkantare) [reply ibm/resource_ibm_container_vpc_cluster_worker_volume_attachment.go:101]: Add one more attribute volume_attachemnt_id to store only the attachment ID

File: ibm/resource_ibm_container_vpc_cluster_worker_volume_attachment.go [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+		return err
+	}
+
+	workersAPI := wpClient.Workers()
+	target, err := getVpcClusterTargetHeader(d, meta)
+	if err != nil {
+		return err
+	}
+
+	clusterNameorID := d.Get("cluster").(string)
```
Comment (hkantare) [inline ibm/resource_ibm_container_vpc_cluster_worker_volume_attachment.go:165]: Don't use d.Get() in Read methods
Import scenrio may not work

File: ibm/resource_ibm_container_vpc_cluster_worker_volume_attachment.go [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+
+	detachVolumeRequest := v2.VolumeRequest{
+		Cluster:            clusterNameorID,
+		VolumeAttachmentID: volumeAttachmentID,
+		Worker:             workerID,
+	}
+
+	_, Err := workersAPI.DeleteStorageAttachment(detachVolumeRequest, target)
+	if err != nil {
+		return fmt.Errorf("Failed to update the the volume attachment to different worker node: %s", Err)
```
Comment (hkantare) [inline ibm/resource_ibm_container_vpc_cluster_worker_volume_attachment.go:203]: Do we need to add any wait logic for delete also?

File: ibm/resource_ibm_container_vpc_cluster_worker_volume_attachment.go [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+	}
+	clusterNameorID := d.Get("cluster").(string)
+	volumeAttachmentID := d.Id()
+	workerID := d.Get("worker").(string)
+
+	createStateConf := &resource.StateChangeConf{
+		Pending: []string{volumeAttaching},
+		Target:  []string{volumeAttached},
+		Refresh: func() (interface{}, string, error) {
+			volume, Err := workersAPI.GetStorageAttachment(clusterNameorID, workerID, volumeAttachmentID, target)
```
Comment (hkantare) [inline ibm/resource_ibm_container_vpc_cluster_worker_volume_attachment.go:256]: Err change to small case

File: ibm/resource_ibm_container_vpc_cluster_worker_volume_attachment.go [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+
+			if Err != nil {
+				return volume, deployInProgress, Err
```
Comment (hkantare) [inline ibm/resource_ibm_container_vpc_cluster_worker_volume_attachment.go:259]: where we declated deployInProgress i think that shd be also in one of the Target

File: examples/ibm-cluster/vpc-cluster-vol-attachment/README.md [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+resource "ibm_container_vpc_worker_storage" "volume_attach"{
+    volume = ibm_is_volume.storage_block.id
+    cluster = ibm_container_vpc_cluster.cluster.id
+    worker = data.ibm_container_vpc_cluster.cluster_info.workers[0]
+}
+
+```
+
+```hcl
+data "iibm_container_vpc_worker_storage" "cluster" {
```
Comment (allen-servedio) [inline examples/ibm-cluster/vpc-cluster-vol-attachment/README.md:95]: Is this right: `iibm_container_vpc_worker_storage` . Shouldn't it be: `ibm_container_vpc_worker_storage`

**Comment by allen-servedio** [comment]
I am curious what the use case is for this? Why not just create a PVC and let Kubernetes create a PV and attach the block storage...?

**Comment by hkantare** [comment]
closing this PR since we have a new PR from Umar

**Pull request closed without merging**